src/util/amount.rs

```diff
@@ -0,0 +1,901 @@
+            if s.chars().skip(1).map(penalty).sum::<i32>() > 1 {
+                return Err(ParseAmountError::InvalidFormat);
+            }
+            return Err(ParseAmountError::Negative);
+        }
+
+        let max_decimals = {
+            // The difference in precision between native (satoshi)
+            // and desired denomination.
+            let precision_diff = -denom.precision();
+            if precision_diff < 0 {
+                // If precision diff is negative, this means we are parsing into a less
+                // precise amount.  That is not allowed unless there are no decimals and
+                // the last digits are zeroes as many as the diffence in precision.
+                let last_n = precision_diff.abs() as usize;
+                if !s.chars().skip(s.len() - last_n).all(|d| d == '0') {
+                    return Err(ParseAmountError::TooPrecise);
+                }
+                s = &s[0..s.len() - last_n];
+                0
+            } else {
+                precision_diff
+            }
+        };
+
+        let mut decimals = None;
+        let mut value: u64 = 0; // as satoshis
+        for c in s.chars() {
+            match c {
+                '0'...'9' => {
+                    // Do `value = 10 * value + digit`, catching overflows.
+                    match 10_u64.checked_mul(value) {
+                        None => return Err(ParseAmountError::TooBig),
+                        Some(val) => match val.checked_add((c as u8 - b'0') as u64) {
+                            None => return Err(ParseAmountError::TooBig),
+                            Some(val) => value = val,
+                        },
+                    }
+                    // Increment the decimal digit counter if past decimal.
+                    decimals = match decimals {
+                        None => None,
+                        Some(d) if d < max_decimals => Some(d + 1),
+                        _ => return Err(ParseAmountError::TooPrecise),
+                    };
+                }
+                '.' => match decimals {
+                    None => decimals = Some(0),
+                    // Double decimal dot.
+                    _ => return Err(ParseAmountError::InvalidFormat),
+                },
+                _ => return Err(ParseAmountError::InvalidFormat),
+            }
+        }
+
+        // Decimally shift left by `max_decimals - decimals`.
+        let scale_factor = max_decimals - decimals.unwrap_or(0);
+        for _ in 0..scale_factor {
+            value = match 10_u64.checked_mul(value) {
+                Some(v) => v,
+                None => return Err(ParseAmountError::TooBig),
+            };
+        }
+
+        Ok(Amount::from_sat(value))
+    }
+
+    /// Express this Amount as a floating-point value in the given denomination.
+    ///
+    /// Please be aware of the risk of using floating-point numbers.
+    pub fn as_float_denom(&self, denom: Denomination) -> f64 {
```

Comment:
Not really sure about `to`/`as`, since you're not really turning this in something else, but just representing the inner value as another type. Anyway, will change. I agree with denomination naming.

Reply:
Hmm, similar to `to_string_in`, I opted for `to_float_in` and `from_str_in`. Do you agree with that?